nexus/src/app/silo.rs

```diff
@@ -140,14 +143,125 @@ impl super::Nexus {
 
     // Users
 
-    pub async fn silo_user_fetch(
+    /// Helper function for looking up a LocalOnly Silo by name
+    ///
+    /// This is called from contexts that are trying to access the "local"
+    /// identity provider.  On failure, it returns a 404 for that identity
+    /// provider.
+    async fn silo_api_fetch(
```

Comment:
I hear that.  I don't love the current name.  Between this and your other comments below, I think there are two separate issues: one is around naming and the other is around these functions being special-purpose for LocalOnly Silos.

On the question of "should these functions be specialized to LocalOnly" silos: the reason is that all of these functions return a 404 saying `not found: identity-provider with name "local"` if you attempt to use them on a Silo that's _not_ LocalOnly.  That's because even though you could, say, fetch a user in a SamlJit Silo, you shouldn't be able to do it through `/system/silos/:silo_name/identity-providers/local` because you don't _have_ a local IdP in that Silo.  I think everything under `../identity-providers/local` ought to produce a 404 in SamlJit Silos.  I suspect we agree on that, but the controversy is maybe that these functions _could_ be general-purpose and maybe _look_ general-purpose _except_ for that error case, and so maybe it feels like that doesn't belong at the Nexus layer?  I wanted to do the validation in the Nexus layer because I worried that exposing general-purpose functions here and relying on the HTTP endpoints to do the extra check creates a pretty big footgun -- it's really easy to forget to do that check and accidentally expose something that we shouldn't.  I think we've mostly adopted the pattern of 1-1 correspondence between HTTP entrypoints and Nexus-level functions so it seemed reasonable.

But here's an idea that might address both the naming and the behavioral concerns?  The root problem here is that although these functions _look_ general-purpose, they're being _used_ via the local IdP.  Maybe that should be more front-and-center.  So instead of:

- `silo_api_fetch` (helper function that just fetches a Silo, but validates that the type is `LocalOnly`) 
- `silo_api_users_list` (lists users for a LocalOnly Silo)
- `silo_api_user_create` (creates a user in a LocalOnly Silo)
- `silo_api_user_delete` (deletes a user from a LocalOnly Silo)
- `silo_api_user_fetch` (fetches a user in a LocalOnly Silo)

Maybe we make these:

- `local_idp_fetch` (helper function that just fetches a Silo, but validates that the type is `LocalOnly`) 
- `local_idp_list_users` (lists users in the local IdP)
- `local_idp_create_user` (creates a user in the local IdP)
- `local_idp_delete_user` (deletes a user from the local IdP)
- `local_idp_fetch_user` (fetches a user in the local IdP)

That way, the idea of the local IdP exists at least somewhat at the Nexus layer and it's enforced there.  The naming and behavior seem less surprising to me.  I wouldn't change the signatures or bodies of these functions from what's here -- just the naming and documentation.  What do you think?

Reply:
> I wanted to do the validation in the Nexus layer because I worried that exposing general-purpose functions here and relying on the HTTP endpoints to do the extra check creates a pretty big footgun -- it's really easy to forget to do that check and accidentally expose something that we shouldn't.

Agree whole heartedly.

> The root problem here is that although these functions look general-purpose, they're being used via the local IdP.

That's my concern, yeah :) Listing and fetching users are general purpose. The only functions unique to LocalOnly silos are creating and deleting users.

I agree that we should be returning 404 for create user and delete user if the Silo's type isn't LocalOnly, but it seems wrong to have IdP specific endpoints for general purpose things. Each IdP's `list_users` is going to be doing the same thing: querying what users are in a Silo and returning SiloUser objects. Each IdP's `fetch_user` is also going to be doing the same thing too.

It's the difference between this:

```
GET /system/silos/:silo_name/identity-providers/local/users/all
POST /system/silos/:silo_name/identity-providers/local/users/create
GET /system/silos/:silo_name/identity-providers/local/users/id/:id
PUT /system/silos/:silo_name/identity-providers/local/users/id/:id
DELETE /system/silos/:silo_name/identity-providers/local/users/id/:id
POST /system/silos/:silo_name/identity-providers/local/users/id/:id/set_password

GET /system/silos/:silo_name/identity-providers/saml/:saml_provider_name/users/all
GET /system/silos/:silo_name/identity-providers/saml/:saml_provider_name/users/id/:id

GET /system/silos/:silo_name/identity-providers/future_idp/:future_idp_provider_name/users/all
GET /system/silos/:silo_name/identity-providers/future_idp/:future_idp_provider_name/users/id/:id
```

and this:

```
GET /system/silos/:silo_name/users/all
GET /system/silos/:silo_name/users/id/:id

POST /system/silos/:silo_name/identity-providers/local/users/create
PUT /system/silos/:silo_name/identity-providers/local/users/id/:id
DELETE /system/silos/:silo_name/identity-providers/local/users/id/:id
POST /system/silos/:silo_name/identity-providers/local/users/id/:id/set_password
```

If we had specific types for SiloUsers based on IdP (LocalOnlySiloUser / SamlJitSiloUser / etc), then it would make more sense to have all those endpoints. How users get created is IdP specific but everything is a SiloUser at the end of the day, and the containing bucket is a Silo.

(I'm realizing now that this should have been RFD feedback instead of PR feedback, sorry).

Reply:
Is your suggestion that the endpoints to list users in a Silo and to fetch an individual user in the Silo get hoisted up to the Silo level instead of putting them at IdP level?  That makes sense to me.  In that case, yeah, the `local_idp_list_users` and `local_idp_fetch_user` functions would no longer be restricted to use with the local IdP and I'd rename them.

There would still be `local_idp_delete_user` and `local_idp_create_user`, and so I would keep the `local_idp_fetch_silo` helper as well.  Does that make sense?

Reply:
Implemented in [5bc7654](https://github.com/oxidecomputer/omicron/pull/1784/commits/5bc7654a53bdd9d90d3343eb8cc71b2786324051).

Reply:
One extra note: I made it `/system/silos/{silo_name}/users/all` and `/system/silos/{silo_name}/users/all` for the reasons that we talked about in the PR.  But I eliminated the extra level of hierarchy in the other place, so it's `POST /system/silos/{silo_name}/identity-providers/local/users` (no trailing `/create`) and `DELETE /system/silos/{silo_name}/identity-providers/local/users/{user_id}` (no extra `id` component).  I didn't see the need for the extra layer there.

Reply:
Nice, resolving this now